### PR TITLE
Pfsync ipv6

### PR DESCRIPTION
**Pull request by lamaral**

### CODE REVIEW

File: sys/netpfil/pf/if_pfsync.c [hunk 1]
```diff
@@ -2436,6 +2643,20 @@ static struct protosw in_pfsync_protosw = {
...
+extern	struct domain inet6domain;
+static struct protosw in6_pfsync_protosw = {
+	.pr_type =		SOCK_RAW,
+	.pr_domain =		&inet6domain,
+	.pr_protocol =		IPPROTO_PFSYNC,
+	.pr_flags =		PR_ATOMIC|PR_ADDR,
+	.pr_input =		pfsync6_input,
+	.pr_output =		rip6_output,
+	.pr_ctloutput =		rip6_ctloutput,
+	.pr_usrreqs =		&rip6_usrreqs
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: Something wrong with indentation here, or is it just GitHub not rendering it properly?

File: sbin/ifconfig/ifpfsync.c [hunk 2]
```diff
@@ -187,24 +217,35 @@ void
...
+		if ((error = getnameinfo(syncpeer_sa, syncpeer_sa->sa_len,
+		    syncpeer, sizeof(syncpeer), NULL, 0, NI_NUMERICHOST)) != 0)
+			errx(1, "getnameinfo: %s", gai_strerror(error));
+		printf("syncpeer: %s ", syncpeer);
+	}
 
 	if (preq.pfsyncr_syncdev[0] != '\0' ||
-	    preq.pfsyncr_syncpeer.s_addr != htonl(INADDR_PFSYNC_GROUP)) {
+	    (preq.pfsyncr_syncpeer.sa.sa_family != AF_UNSPEC &&
+		preq.pfsyncr_syncpeer.in4.sin_addr.s_addr != htonl(INADDR_PFSYNC_GROUP))) {
```
Comment (TuxPowered42) [inline sbin/ifconfig/ifpfsync.c]: Printing of status for IPv6 is missing.

File: sys/netpfil/pf/if_pfsync.c [hunk 3]
```diff
@@ -90,28 +90,43 @@ __FBSDID("$FreeBSD$");
...
-	struct in_addr src;
+#ifdef INET
+	struct ip *ip4;
+	struct in_addr src4;
+#endif
+#ifdef INET6
+	struct ip6_hdr *ip6;
+	struct in6_addr src6;
+#endif
+	sa_family_t af;
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: The IPv4 and IPv6 parts could be made into an union.

File: sys/netpfil/pf/if_pfsync.c [hunk 4]
```diff
@@ -689,8 +707,118 @@ pfsync_input(struct mbuf **mp, int *offp __unused, int proto __unused)
 	}
 
 	/* Cheaper to grab this now than having to mess with mbufs later */
-	pkt.ip = ip;
-	pkt.src = ip->ip_src;
+	pkt.af = AF_INET6;
+	pkt.ip4 = ip;
+	pkt.src4 = ip->ip_src;
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: The family is AF_INET6 but the IPv4 fields are filled in.

Comment (TuxPowered42) [reply sys/netpfil/pf/if_pfsync.c]: The whole function could be `#ifdef AF_INET`. I can't leave a comment on non-changed line, though.

File: sys/netpfil/pf/if_pfsync.c [hunk 5]
```diff
@@ -689,8 +707,118 @@ pfsync_input(struct mbuf **mp, int *offp __unused, int proto __unused)
...
+	}
+	PF_RULES_RUNLOCK();
+
+done:
+	m_freem(m);
+	return (IPPROTO_DONE);
+}
+
+static int
+pfsync6_input(struct mbuf **mp, int *offp __unused, int proto __unused)
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: The whole function cold be `#ifdef INET6`.

File: sys/netpfil/pf/if_pfsync.c [hunk 6]
```diff
@@ -1384,19 +1511,20 @@ pfsyncioctl(struct ifnet *ifp, u_long cmd, caddr_t data)
 			sifp = NULL;
 		else if ((sifp = ifunit_ref(pfsyncr.pfsyncr_syncdev)) == NULL)
 			return (EINVAL);
-
-		if (sifp != NULL && (
-		    pfsyncr.pfsyncr_syncpeer.s_addr == 0 ||
-		    pfsyncr.pfsyncr_syncpeer.s_addr ==
-		    htonl(INADDR_PFSYNC_GROUP)))
+		if (sifp != NULL && (sc->sc_sync_peer.sa.sa_family == AF_UNSPEC)) {
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: This forces IPv4 multicast as default when IP address is not given. I'd love to see IPv6 instead.

File: sys/netpfil/pf/if_pfsync.c [hunk 7]
```diff
@@ -1384,19 +1511,20 @@ pfsyncioctl(struct ifnet *ifp, u_long cmd, caddr_t data)
...
+		}
 
 		PFSYNC_LOCK(sc);
-		if (pfsyncr.pfsyncr_syncpeer.s_addr == 0)
-			sc->sc_sync_peer.s_addr = htonl(INADDR_PFSYNC_GROUP);
-		else
-			sc->sc_sync_peer.s_addr =
-			    pfsyncr.pfsyncr_syncpeer.s_addr;
+		printf("pfsync: Address family: %d\n", sc->sc_sync_peer.sa.sa_family);
+		if (sifp != NULL && (sc->sc_sync_peer.sa.sa_family == AF_UNSPEC)) {
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: It's the same condition as above. You could have  just one `if` and merge those 2 code blocks. Just watch out for `PFSYNC_LOCK`.

File: sys/netpfil/pf/if_pfsync.c [hunk 8]
```diff
@@ -1442,17 +1570,41 @@ pfsyncioctl(struct ifnet *ifp, u_long cmd, caddr_t data)
...
-		ip->ip_tos = IPTOS_LOWDELAY;
-		/* len and id are set later. */
-		ip->ip_off = htons(IP_DF);
-		ip->ip_ttl = PFSYNC_DFLTTL;
-		ip->ip_p = IPPROTO_PFSYNC;
-		ip->ip_src.s_addr = INADDR_ANY;
-		ip->ip_dst.s_addr = sc->sc_sync_peer.s_addr;
+		printf("pfsync: Before template. AF: %d\n", sc->sc_sync_peer.sa.sa_family);
+		switch (sc->sc_sync_peer.sa.sa_family) {
+		case AF_INET: {
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: Please add `ifdef …` for address families.

File: sys/netpfil/pf/if_pfsync.c [hunk 9]
```diff
@@ -1589,20 +1740,41 @@ pfsync_sendout(int schedswi, int c)
 	m->m_len = m->m_pkthdr.len = b->b_len;
 
 	/* build the ip header */
-	ip = (struct ip *)m->m_data;
-	bcopy(&sc->sc_template, ip, sizeof(*ip));
-	offset = sizeof(*ip);
+	switch (sc->sc_sync_peer.sa.sa_family) {
+	case AF_INET:
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: Please add `ifdef …` for address families.

File: sys/netpfil/pf/if_pfsync.c [hunk 10]
```diff
@@ -1831,7 +2003,10 @@ pfsync_defer_tmo(void *arg)
 		free(pd, M_PFSYNC);
 	PFSYNC_BUCKET_UNLOCK(b);
 
-	ip_output(m, NULL, NULL, 0, NULL, NULL);
+	if (sc->sc_sync_peer.sa.sa_family == AF_INET)
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: Please add `ifdef …` for address families. You might need to use a `switch` expression for that.

File: sys/netpfil/pf/if_pfsync.c [hunk 11]
```diff
@@ -2347,10 +2522,29 @@ pfsyncintr(void *arg)
 			 * own pfsync packet based on M_SKIP_FIREWALL
 			 * flag. This is XXX.
 			 */
-			if (m->m_flags & M_SKIP_FIREWALL)
-				ip_output(m, NULL, NULL, 0, NULL, NULL);
-			else if (ip_output(m, NULL, NULL, IP_RAWOUTPUT, &sc->sc_imo,
-			    NULL) == 0)
+			if (m->m_flags & M_SKIP_FIREWALL) {
+				printf("pfsync: transmit packet skip\n");
+				if (sc->sc_sync_peer.sa.sa_family == AF_INET)
```
Comment (TuxPowered42) [inline sys/netpfil/pf/if_pfsync.c]: Please add `ifdef …` for address families.

**Review by TuxPowered42 — CHANGES_REQUESTED** [review]

File: sys/netpfil/pf/if_pfsync.c (hunk 1 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: GitHub doesn't deal well with tabs. On the editor it looks good.

File: sys/netpfil/pf/if_pfsync.c (hunk 8 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: Done.

File: sys/netpfil/pf/if_pfsync.c (hunk 9 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: Done.

File: sys/netpfil/pf/if_pfsync.c (hunk 10 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: Done

File: sys/netpfil/pf/if_pfsync.c (hunk 11 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: Done

File: sys/netpfil/pf/if_pfsync.c (hunk 6 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: I'm not sure how happy the devs would be to change the default behavior of using IPv4.
We could maybe make a sysctl to allow for choosing it?

File: sys/netpfil/pf/if_pfsync.c (hunk 4 above)
Comment (lamaral) [reply sys/netpfil/pf/if_pfsync.c]: This function is `pfsync_input()` not `pfsync6_input()`. This is the IPv4 version.

**Pull request closed without merging**